homeassistant/components/camera/__init__.py

```diff
@@ -0,0 +1,1133 @@
+                    camera.stop_recording()
+
+            elif action == SWITCH_ACTION_SNAPSHOT:
+                if (state == STATE_ON and not
+                        camera.is_taking_snapshot):
+                    camera.take_snapshot()
+
+            camera.update_ha_state(True)
+
+    hass.services.register(
+        DOMAIN,
+        SERVICE_CAMERA,
+        handle_motion_detection_service)
+
+    return True
+
+
+# pylint: disable=too-many-instance-attributes
+# pylint: disable=too-many-public-methods
+class Camera(Entity):
+    """
+    Base class for cameras.
+    This is quite a large class but the camera component encompasses a lot of
+    functionality.  It should take care of most of the heavy lifting and
+    plumbing associated with adding support for additional models of camera.
+    If you are adding support for a new camera your entity class should inherit
+    from this.
+    """
+
+    def __init__(self, hass, device_info):
+        self.hass = hass
+        self.device_info = device_info
+        self.base_url = device_info.get('base_url')
+        if not self.base_url.endswith('/'):
+            self.base_url = self.base_url + '/'
+        self.username = device_info.get('username')
+        self.password = device_info.get('password')
+        self.is_streaming = False
+        self._is_detecting_motion = False
+        self._last_motion_detected = datetime.datetime.now()
+        self.last_connected_address = None
+        self._still_image_url = device_info.get('still_image_url', 'image.jpg')
+        # these are the camera functions and capabilities initialised
+        # to defaults, these should be overridden in derived classes
+        self._is_motion_detection_supported = False
+        self._is_motion_detection_enabled = False
+
+        self._is_recording = False
+        self._is_taking_snapshot = False
+
+        self._is_ftp_upload_supported = False
+        self._is_ftp_upload_enabled = False
+        self._can_auto_configure_ftp = False
+
+        self._is_ftp_configured = False
+        self._ftp_host = ''
+        self._ftp_port = 21
+        self._ftp_username = ''
+        self._ftp_password = ''
+        self._ftp_relative_path = ''
+
+        self._logger = logging.getLogger(__name__)
+
+        self._images_path = device_info.get('images_path', None)
+
+        if (self._images_path is not None and not
+                os.path.isdir(self._images_path)):
+            os.makedirs(self._images_path)
+
+        self._event_images_path = None
+        self._recording_images_path = None
+        self._snapshot_images_path = None
+        self._ftp_path = None
+
+        self._child_entities = {}
+
+        self._action_entities = {}
+
+        self._ftp_configurator = None
+
+        self._lastconfig_update_from_device = None
+
+        self.hass.bus.listen(
+            EVENT_FTP_FILE_RECEIVED,
+            self.process_file_event)
+
+    def add_child_component_listeners(self):
```

Comment:
Would it be possible to extract the child component logic and data and put it into a separate class which is kept in `self._action_entities` or `self._child_entities`? It wouldn't matter if each class is only instantiated once for each object as it would help organize different sections of logic together.